### DISCUSSION TITLE
Properly Removing Corrupted or Unwanted Versions from the Repository

### DISCUSSION THREAD

**Discussion by its-monotype** [dennisvang/tufup#117] [Q&A] [answered]
I would like to know the correct steps for removing corrupted or unwanted versions from the repository. Here's what I assume should be done:

1. Modify `targets.json` metadata:
   - Remove the entries related to the specific version from the `signed.targets` object. This includes removing the `.tar.gz` and `.patch` references that correspond to the version we want to remove.

2. Remove files from the repository:
   - Delete the actual `.tar.gz` and `.patch` files from the repository's `targets` directory.

Could someone please confirm if my understanding is correct? Are there any additional steps or precautions I should be aware of when performing this task?

**Comment by dennisvang** [comment]
@its-monotype Good question.

In addition to the steps described above, you'll also need to re-sign the affected metadata files.

Note that `tufup` provides a shortcut for the removal of the *latest* target. This removes the target files and updates the metadata files accordingly, including re-signing.

From the command line:

```
tufup targets remove-latest <path-to-keystore>
```

Also see CLI help: `tufup targets -h`

The `remove-latest` CLI implementation can be used as an example for scripting. As you can see [here][1], it uses [`Repository.remove_latest_bundle()`][2] followed by `Repository.publish_changes()`.

If you want to remove a specific version that is not the latest, you can use [`Repository.remove_latest_bundle()`][2] as an example.

Alternatively, you could remove the metadata entries manually, then re-sign.
 

[1]: https://github.com/dennisvang/tufup/blob/6d2c707adc66fd7637739f31f9859b0298a4053d/src/tufup/repo/cli.py#L279
[2]: https://github.com/dennisvang/tufup/blob/6d2c707adc66fd7637739f31f9859b0298a4053d/src/tufup/repo/__init__.py#L747

**Comment by RasmusN** [comment]
Almost the same question:
Is it safe to delete old versions from the `targets` directory in the remote repository? 

If so, can I just delete them or do I need to update some metadata files?

**Comment by dennisvang** [comment, marked as answer]
> Is it safe to delete old versions from the `targets` directory in the remote repository?

@RasmusN Good question.

Removing old versions, i.e. versions other than the latest version, is only safe if you use *full* updates.

Presently, `tufup` does not check if the current (running) version of your app is available in the targets list, so *full* updates from any old version to the latest one should still work.

However, removing old versions is *not* safe if you use *patch* updates, because these depend upon each other. Removing a patch, other than the latest one, will break (part of) the chain. Currently, there is no safeguard against this. That's why the current implementation only has a built-in command for rolling back the *latest* version (`remove-latest`).

Beware that patch updates are [enabled by default][2], but you can disable them using the `patch` argument: 

```python
check_for_updates(patch=False)
```

> If so, can I just delete them or do I need to update some metadata files?

In principle, you can remove any version, as long as the change is reflected in the metadata. Have a look at the `remove-latest` command, as [described above][1], to see what this entails.

Basically, if you remove and/or modify any file from the targets directory, the `targets.json` file must be updated to reflect the new state of that directory. If `targets.json` is updated, then `snapshot.json` and `timestamp.json` must also be updated. If you update targets metadata using the built-in methods, snapshot and timestamp metadata will be updated automatically.

[1]: https://github.com/dennisvang/tufup/issues/74#issuecomment-1653774427
[2]: https://github.com/dennisvang/tufup/blob/6d2c707adc66fd7637739f31f9859b0298a4053d/src/tufup/client.py#L136
